src/core/buffer/index.ts

```diff
@@ -401,29 +400,52 @@ function Buffer({
         .map((args) => objectAssign({ segment }, args));
     }
 
-    function onClockTick(
-      [ timing, _wantedBufferAhead, _maxBufferBehind, _maxBufferAhead ]
-        : [ IBufferClockTick, number, number, number ],
-      i : number
+    /**
+     * Check and load needed segments.
+     * @param {Object} timing
+     * @param {number} _wantedBufferAhead
+     * @param {number} _maxBufferAhead
+     * @param {Boolean} needsInitSegment
+     * @returns {Observable}
+     */
+    function checkNeededSegments(
+      timing : IBufferClockTick,
+      _wantedBufferAhead : number,
+      _maxBufferAhead : number,
+      needsInitSegment : boolean
     ) : Observable<{ segment : Segment } & IDownloaderResponse> {
       const bufferGoal = Math.min(_wantedBufferAhead, _maxBufferAhead);
-      const loadNeeded$ = getNeededSegments(timing, bufferGoal, i)
+      return getNeededSegments(timing, bufferGoal, needsInitSegment)
         .concatMap(loadNeededSegments);
-      const clean$ = cleanBuffer(
-        bufferingQueue, timing.currentTime, _maxBufferBehind, _maxBufferAhead);
-      return Observable.merge(loadNeeded$, clean$);
     }
 
-    const segmentsPipeline = Observable.combineLatest(
-      clock$,
-      wantedBufferAhead,
-      maxBufferBehind,
-      maxBufferAhead
-    )
-      .mergeMap(onClockTick)
+    /**
+     * Check the size of buffer and clean if it goes beyond the limits set by
+     * the user.
+     * @param {number} currentTime
+     * @param {number} _maxBufferBehind
+     * @param {number} _maxBufferAhead
+     * @returns {Observable}
+     */
+    function checkBufferSize(
+      currentTime : number,
+      _maxBufferBehind : number,
+      _maxBufferAhead : number
+    ) : Observable<never> {
+      return cleanBuffer(
+        bufferingQueue, currentTime, _maxBufferBehind, _maxBufferAhead);
+    }
+
+    const loadNeededSegments$ = Observable
+      .combineLatest(clock$, wantedBufferAhead, maxBufferAhead)
+      .concatMap((args, i) => checkNeededSegments(args[0], args[1], args[2], !i))
       .concatMap(appendDataInBuffer);
 
-    return Observable.merge(segmentsPipeline, messageSubject)
+    const clean$ = Observable
+      .combineLatest(clock$, maxBufferBehind, maxBufferAhead)
+      .mergeMap(args => checkBufferSize(args[0].currentTime, args[1], args[2]));
+
+    return Observable.merge(loadNeededSegments$, clean$, messageSubject)
```

Comment:
The cleaning + downloading tasks where previously done both in ``onClockTick``, called by a mergeMap.

Since mergeMap has been updated to switchMap, cleaning of the buffer would not be done until the segments from the previous tick have been downloaded.

To resolve that issue we do both the concatMap thing for segments, and the mergeMap thing for cleaning when the clock ticks.

This read that way:
  - add segments to download in the queue if needed on each clock tick
  - see and clean superfluous part of the buffer on clock ticks

instead of just, in the case of the concatMap:
  - add segments to download + cleaning task in the queue on each clock tick.